core/src/main/java/com/linecorp/armeria/client/cookie/DefaultCookieJar.java

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.cookie;
+
+import static java.util.Objects.requireNonNull;
+
+import java.net.URI;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.locks.ReentrantLock;
+
+import javax.annotation.Nullable;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Strings;
+
+import com.linecorp.armeria.common.Cookie;
+import com.linecorp.armeria.common.CookieBuilder;
+import com.linecorp.armeria.common.Cookies;
+
+/**
+ * A default in-memory {@link CookieJar} implementation.
+ */
+final class DefaultCookieJar implements CookieJar {
+
+    private final Set<Cookie> store;
+    /**
+     * Used to find cookies for a host that matches a domain. For example, if there is a domain example.com,
+     * host example.com or foo.example.com will match all cookies in that entry.
+     */
+    private final Map<String, Set<Cookie>> filter;
+    private final CookiePolicy cookiePolicy;
+    private final ReentrantLock lock;
+
+    DefaultCookieJar() {
+        this(CookiePolicy.acceptOriginalServer());
+    }
+
+    DefaultCookieJar(CookiePolicy cookiePolicy) {
+        this.cookiePolicy = cookiePolicy;
+        store = new HashSet<>();
```

Comment:
curious do we need `store`, looks like it's possible to just use filter in current methods?

Reply:
I think a store helps when removing cookie because we don't have to go through the filter and do domain matching again to find where it is.